otherlibs/stdune/fpath.ml

```diff
@@ -14,35 +15,60 @@ let mkdir ?(perms = 0o777) t_s =
   with
   | Unix.Unix_error (EEXIST, _, _) -> Already_exists
   | Unix.Unix_error (ENOENT, _, _) -> Missing_parent_directory
+  | Unix.Unix_error (ENOTDIR, _, _) -> Parent_not_directory
 
 type mkdir_p_result =
   | Already_exists
   | Created
+  | Already_exists_not_directory of string
+
+let dyn_of_mkdir_p_result = function
+  | Already_exists -> Dyn.Variant ("Already_exists", [])
+  | Already_exists_not_directory path ->
+    Dyn.Variant ("Already_exists_not_directory", [ Dyn.string path ])
+  | Created -> Dyn.Variant ("Created", [])
 
 let rec mkdir_p ?perms t_s =
   match mkdir ?perms t_s with
   | Created -> Created
-  | Already_exists -> Already_exists
+  | Already_exists ->
+    if Stdlib.Sys.is_directory t_s then Already_exists
```

Comment:
This seems pretty bad to me performance-wise. We call `mkdir_p` a lot, and in most cases, it returns the `Already_exists` variant because we've already created the directory. Your change makes us waste one extra system call in these cases.

Reply:
I was also suspicious about the performance here and I forgot to ask about it. Sorry about that.